### PR TITLE
Compile functions

### PR DESCRIPTION
**Pull request by rbartlensky**
At the moment function calls are not supported. In the following PRs the compiler and vm will support functions calls, call arguments, and return values.

### CODE REVIEW

File: luacompiler/src/lib/bytecode/instructions.rs [hunk 1]
```diff
@@ -40,6 +40,7 @@ pub fn format_instr(instr: u32) -> String {
         10 => "Exp",
         11 => "GetAttr",
         12 => "SetAttr",
+        13 => "Closure",
```
Comment (ltratt) [inline luacompiler/src/lib/bytecode/instructions.rs]: Why are we using the term "closure" whereas I thought there were only functions?

**Comment by ltratt** [comment]
At first glance, this looks mostly OK, but we absolutely need source-level tests of this stuff too. So I would expect tests that look something like:

```rust
#[test]
fn t() {
    assert_eq!(compile_and_run("function f(x) return x + 1 end; print(f(2))), "2");
}
```

[I compressed that down a bit, but hopefully you get the idea.]

It would also be fine if the Lua tests were in their own source files, a bit like rustc does things.

File: luacompiler/src/lib/bytecode/instructions.rs (hunk 1 above)
Comment (rbartlensky) [reply luacompiler/src/lib/bytecode/instructions.rs]: Functions can also modify local variables that were defined in outer scopes. From this point of view, they are closures. Also, the lua implementation translates functions to closures.

Comment (ltratt) [reply luacompiler/src/lib/bytecode/instructions.rs]: OK. [I'm not sure the terminological shift is useful, but it doesn't hurt much either. In most languages, I see functions and closures as synonyms from a normal user's perspective.]

**Comment by rbartlensky** [comment]
Is this fine for now? I can't print anything to stdout yet, but I will be able to do so in 1-2 PRs.

**Comment by ltratt** [comment]
Doesn't each test case need to have a check (of some sort) that determines whether it passed successfully or not?

**Comment by rbartlensky** [comment]
I am not sure what kind of checks I can have at this point. At the moment my check is simply: "does it crash or not?"

I have a few unit-tests in the [vm](https://github.com/rbartlensky/Lua-interpreter/blob/compile-functions/luavm/src/lib/mod.rs#L86-L107), which test the state of the registers after a program is interpreted. In my next PR, I will also have function calls, and for that I have added extra test cases. Once I implement the `print` function I will add tests which check the output of the programs.

**Comment by ltratt** [comment]
OK, please squash.

**Comment by rbartlensky** [comment]
I think I will leave the commits as they are, done!

**Pull request merged**